### PR TITLE
Feat: Improve ingestion speed by running more routinues

### PR DESCRIPTION
**Pull request by StrongMonkey**
We should be running more gorountine to parse and create embedding for documents to speed up ingestion time. Mostly of operation are not heavily relied on CPU. Specially when creating embedding with documents, we don't need to constraint with the number of core system has because we are just waiting for OPEN AI api calls and are not spending cpu/io resource locally.

This also relies on https://github.com/iwilltry42/langchaingo/pull/1.

Tested with new changes and it speed up ingestion time from 400 seconds down to 38 seconds for a 3000 page PDF.

Will re-run e2e tests but this should not impact our ingestion quality.

### CODE REVIEW

File: pkg/vectorstore/chromem/chromem.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func (s *Store) AddDocuments(ctx context.Context, docs []vs.Document, collection
 		return nil, vs.ErrCollectionNotFound{Collection: collection}
 	}
 
-	err := col.AddDocuments(ctx, chromemDocs, runtime.NumCPU()/2)
+	err := col.AddDocuments(ctx, chromemDocs, int(math.Ceil(float64(len(chromemDocs))/20.0)))
```
Comment (iwilltry42) [inline pkg/vectorstore/chromem/chromem.go:61]: Can we make this configurable?

**Review by iwilltry42 — CHANGES_REQUESTED** [review]

File: pkg/vectorstore/chromem/chromem.go (hunk 1 above)
Comment (StrongMonkey) [reply pkg/vectorstore/chromem/chromem.go:61]: Good idea! will do

File: pkg/datastore/documentloader/pdf.go [hunk 2]
```diff
@@ -41,6 +44,7 @@ type PDF struct {
 	opts      PDFOptions
 	document  *fitz.Document
 	converter *md.Converter
+	lock      sync.Mutex
```
[the remark this answers is not in the document]
Comment (StrongMonkey) [reply pkg/datastore/documentloader/pdf.go]: Addressed.

File: pkg/datastore/documentloader/pdf.go [hunk 3]
```diff
@@ -67,47 +71,54 @@ func NewPDF(r io.Reader, optFns ...func(o *PDFOptions)) (*PDF, error) {
...
+				Content: strings.TrimSpace(markdown),
+				Metadata: map[string]any{
+					"page":       pageNum + 1,
+					"totalPages": numPages,
+				},
+			}
+			l.lock.Lock()
+			docs = append(docs, doc)
+			l.lock.Unlock()
+			return nil
```
[the remark this answers is not in the document]
Comment (StrongMonkey) [reply pkg/datastore/documentloader/pdf.go]: Addressed 👍

File: pkg/vectorstore/chromem/chromem.go (hunk 1 above)
Comment (StrongMonkey) [reply pkg/vectorstore/chromem/chromem.go]: Addressed as an environment variable. Also, I make this a fix value to be the number of goroutinues for user to understand it better.

File: pkg/datastore/documentloader/pdf.go [hunk 4]
```diff
@@ -67,47 +74,59 @@ func NewPDF(r io.Reader, optFns ...func(o *PDFOptions)) (*PDF, error) {
...
 }
 
 // Load loads the PDF document and returns a slice of vs.Document containing the page contents and metadata.
 func (l *PDF) Load(ctx context.Context) ([]vs.Document, error) {
 	docs := make([]vs.Document, 0, l.document.NumPage())
+	numPages := l.document.NumPage()
 
-	for pageNum := 0; pageNum < l.document.NumPage(); pageNum++ {
+	g, childCtx := errgroup.WithContext(ctx)
+	g.SetLimit(env.GetIntFromEnvOrDefault(PDF_LOAD_PARALLEL_THREAD, 100))
```
Comment (iwilltry42) [inline pkg/datastore/documentloader/pdf.go:87]: Looks quite pythonic :D

**Review by njhale — APPROVED** [review]
LGTM as soon as the replaces directive is updated!

File: pkg/datastore/documentloader/pdf.go [hunk 5]
```diff
@@ -4,14 +4,20 @@ import (
...
 	"github.com/PuerkitoBio/goquery"
 	"github.com/gen2brain/go-fitz"
 	"github.com/gptscript-ai/knowledge/pkg/datastore/types"
+	"github.com/gptscript-ai/knowledge/pkg/env"
 	vs "github.com/gptscript-ai/knowledge/pkg/vectorstore"
+	"golang.org/x/sync/errgroup"
 )
 
+// PDF_LOAD_PARALLEL_THREAD can be set as an environment variable that controls the number of goroutines to load pdf pages
+const PDF_LOAD_PARALLEL_THREAD = "PDF_LOAD_PARALLEL_THREAD"
```
Comment (iwilltry42) [inline pkg/datastore/documentloader/pdf.go:19]: Maybe something more "reusable" like `DOCUMENTLOADER_<Loader Name>_<Setting>` ?

E.g. `DOCUMENTLOADER_PDF_NUM_THREADS` ?

Comment (iwilltry42) [reply pkg/datastore/documentloader/pdf.go:19]: Or maybe even as a PDF option so one can set it via config?

File: pkg/datastore/documentloader/pdf.go (hunk 4 above)
Comment (StrongMonkey) [reply pkg/datastore/documentloader/pdf.go:87]: Hmm yeah, will change the name 👍

File: pkg/datastore/documentloader/pdf.go (hunk 5 above)
Comment (StrongMonkey) [reply pkg/datastore/documentloader/pdf.go:19]: yeah, I think putting into pdf option is a good way. Will address 👍

File: pkg/vectorstore/chromem/chromem.go [hunk 6]
```diff
@@ -4,14 +4,17 @@ import (
...
 	"strconv"
 
 	"github.com/google/uuid"
+	"github.com/gptscript-ai/knowledge/pkg/env"
 	vs "github.com/gptscript-ai/knowledge/pkg/vectorstore"
 	"github.com/philippgille/chromem-go"
 )
 
+// EmbeddingParallelThread can be set as an environment variable to control the number of parallel API calls to create embedding for documents. Default is 100
+const EmbeddingParallelThread = "EMBEDDING_PARALLEL_THREAD"
```
Comment (iwilltry42) [inline pkg/vectorstore/chromem/chromem.go:16]: Last one, I promise :)

```suggestion
const VsChromemEmbeddingParallelThread = "VS_CHROMEM_EMBEDDING_PARALLEL_THREAD"
```

**Review by iwilltry42 — CHANGES_REQUESTED** [review]

File: pkg/vectorstore/chromem/chromem.go (hunk 6 above)
Comment (StrongMonkey) [reply pkg/vectorstore/chromem/chromem.go]: Addressed!

**Comment by StrongMonkey** [comment]
@iwilltry42 Also, we need to have this PR merged first https://github.com/iwilltry42/langchaingo/pull/1. Not sure what you are planning on merging that on upstream

**Comment by iwilltry42** [comment]
@StrongMonkey once I'm back, I'll create another branch and change bases, so I can have my upstream PR and both of our changes in another branch 👍

**Comment by iwilltry42** [comment]
Or faster - let's use your fork for now 🤔

**Comment by StrongMonkey** [comment]
Ok... once you approve, I will change the go.mod to point to my branch(for at least now)

**Review by iwilltry42 — APPROVED** [review]

**Pull request merged**